### ISSUE TITLE
File manifests encode empty link names.

### ISSUE THREAD

**Issue by Jorropo** [ipfs/go-ipfs#8691]
### Checklist

- [X] This is a bug report, not a question. Ask questions on [discuss.ipfs.io](https://discuss.ipfs.io).
- [X] I have searched on the [issue tracker](https://github.com/ipfs/go-ipfs/issues?q=is%3Aissue) for my bug.
- [X] I am running the latest [go-ipfs version](https://dist.ipfs.io/#go-ipfs) or have an issue updating.

### Installation method

built from source

### Version

```Text
$ ipfs version --all
go-ipfs version: 0.12.0-rc1-8a88718ea
Repo version: 12
System version: amd64/linux
Golang version: go1.17.5
```

### Config

```json
N/A
```

### Description

I'm implementing a very low copy chunking + layout engine exporting `.car` and wondering why my CIDs are different.
This is because my IPLD nodes don't contain names when not needed.
This file for example:
Mine (expected):
```
$ ipfs block get bafybeig76jnqhurgk4uh57a7soykyi2pcr673tfiw72uvsp5neexgxnp2i | protoc --decode=PBNode ipld.pb
Data: "\010\002\030\222N \2000 \222\036"
Links {
  Hash: "\001U\022 \262\307\016\013 77\r\204\370\260\332\315@\245\201[\375\243\251\237\232\'\013\364\272\314\305\213\226\315\310"
  Tsize: 6144
}
Links {
  Hash: "\001U\022 \311\254\220r?,\276\'\345\034\276v\311\322\177\321`\016\233\210\n<\3566\250*\365g\372\220l\304"
  Tsize: 3858
}
```
IPFS (`#` comments added by me):
```
ipfs block get bafybeibpwnfexq45cqgf27w7wyzy2ebzgizwieushxt6jidzukyirnxss4 | protoc --decode=PBNode ipld.pb
Data: "\010\002\030\222N \2000 \222\036"
Links {
  Hash: "\001U\022 \262\307\016\013 77\r\204\370\260\332\315@\245\201[\375\243\251\237\232\'\013\364\272\314\305\213\226\315\310"
  Name: "" # <- EXTRA HERE
  Tsize: 6144
}
Links {
  Hash: "\001U\022 \311\254\220r?,\276\'\345\034\276v\311\322\177\321`\016\233\210\n<\3566\250*\365g\372\220l\304"
  Name: "" # <- EXTRA HERE
  Tsize: 3858
}
```
Both files load and work perfectly.
FYI, this a two leaf file with raw-leaves and that the root block.
This cost us 2 bytes per in-file-link.
Everything else (the leaves and unixfs's data match perfectly).

I know this is not really an IPFS bug but more an IPLD one.
However this is not unixfs since this is on the IPLD node object, and I don't know yet which code deals with that, so I'm reporting it and will link a fix PR to this issue.

### Breaking hashes ?
I belive it would be nice to bundle a fix for this in the yet to be scheduled "big hash break update" (I think that how we should name such update).

**Comment by Jorropo** [comment]
Discussed on the IPLD comunity call, it seems the new `go-ipld-prime` code and IPLD specs do are on the side of `Name` should be omited here.

So this might be a hash convergence "bug" while moving to that.

**Comment by rvagg** [comment]
I believe this is is to blame for the current state where we get a blank `Name` regardless of what's been set: https://github.com/ipfs/go-merkledag/blob/437cdd83c8e85406ba782bb456bb654d09ec3d27/coding.go#L120, this is not a pointer: https://github.com/ipfs/go-ipld-format/blob/7fd282bb7e4a59d4c544526fad89670d6259e7f4/format.go#L50, so it's going to have a default `""`.

I'm not aware of anything in our stack that prevents `Name` from being omitted, all our code treats it as optional and we therefore [spec'ed it that way](https://ipld.io/specs/codecs/dag-pb/spec/#logical-format). If we're intending on doing a hash-breaking change at some point this would be a good candidate. Maybe this could be coupled with a push to use go-ipld-prime for unixfs (via https://github.com/ipfs/go-unixfsnode)?

**Comment by Jorropo** [comment]
this is not a pointer: https://github.com/ipfs/go-ipld-format/blob/7fd282bb7e4a59d4c544526fad89670d6259e7f4/format.go#L50, so it's going to have a default "".
@rvagg 

I don't belive so, go-pls tell me it is actually, which **do is a pointer**:
https://github.com/ipfs/go-merkledag/blob/437cdd83c8e85406ba782bb456bb654d09ec3d27/pb/merkledag.pb.go#L44
See the `&` here:
```go
pbn.Links[i].Name = &l.Name
```
Changing it to:
```go
if len(l.Name) > 0 {
  pbn.Links[i].Name = &l.Name
}
```
Should be enough to fix it no ?

**Comment by rvagg** [comment]
yeah, that should be enough; will have to defer to @aschmahmann wrt breaking hash compatibility though, it's a pretty big deal and this would have wide reaching impact—maybe too much to make it worth saving a couple of bytes (as I mentioned on the call the other day, dag-pb is mostly history and we try not to touch it too much anymore!).

what I meant was that Link.Name in go-ipld-format isn't a pointer, but then a pointer to its value gets used in go-merkledag, which is where the `""` is coming from

**Comment by Jorropo** [comment]
BTW, I've just tried this fix:
```go
if len(l.Name) > 0 {
  pbn.Links[i].Name = &l.Name
}
```
And doesn't work, it still produce `Name: ""` fields.
